kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/LikeExpressionEvaluator.java

```diff
@@ -0,0 +1,171 @@
+import io.delta.kernel.types.BooleanType;
+import io.delta.kernel.types.DataType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.internal.util.Utils;
+
+import static io.delta.kernel.defaults.internal.DefaultEngineErrors.invalidEscapeSequence;
+import static io.delta.kernel.defaults.internal.DefaultEngineErrors.unsupportedExpressionException;
+
+/**
+ * Utility methods to evaluate {@code like} expression.
+ */
+public class LikeExpressionEvaluator {
+    private LikeExpressionEvaluator() {
+    }
+
+    static Predicate validateAndTransform(
+            Predicate like,
+            List<Expression> childrenExpressions, List<DataType> childrenOutputTypes) {
+        int size = childrenExpressions.size();
+        if (size < 2 || size > 3) {
+            throw unsupportedExpressionException(like,
+                    "Invalid number of inputs to LIKE expression. " +
+                            "Example usage: LIKE(column, 'test%'), LIKE(column, 'test\\[%', '\\')");
+        }
+
+        Expression left = childrenExpressions.get(0);
+        DataType leftOutputType = childrenOutputTypes.get(0);
+        Expression right = childrenExpressions.get(1);
+        DataType rightOutputType = childrenOutputTypes.get(1);
+        Expression escapeCharExpr = size==3 ? childrenExpressions.get(2) : null;
+        DataType escapeCharOutputType = size==3 ? childrenOutputTypes.get(2) : null;
+
+        if (!(StringType.STRING.equivalent(leftOutputType)
+                && StringType.STRING.equivalent(rightOutputType))) {
+            throw unsupportedExpressionException(like,
+                    "LIKE is only supported for string type expressions");
+        }
+
+        if (escapeCharExpr != null &&
+                (!(escapeCharExpr instanceof Literal &&
+                StringType.STRING.equivalent(escapeCharOutputType)))) {
+            throw unsupportedExpressionException(like,
+                    "LIKE expects escape token expression to be a literal of String type");
+        }
+
+        Literal literal = (Literal) escapeCharExpr;
+        if (literal!=null &&
+                literal.getValue().toString().length() != 1) {
+            throw unsupportedExpressionException(like,
+                    "LIKE expects escape token to be a single character");
+        }
+
+        return new Predicate(like.getName(), Arrays.asList(left, right, escapeCharExpr));
+    }
+
+    static ColumnVector eval(List<ColumnVector> children) {
+        final char DEFAULT_ESCAPE_CHAR = '\\';
+
+        return new ColumnVector() {
+            final ColumnVector escapeCharLiteral =
+                    children.size() == 3 ?
+                    children.get(2) :
+                    null;
+            final ColumnVector left = children.get(0);
+            final ColumnVector right = children.get(1);
+
+            final char escape = escapeCharLiteral!=null ?
+                    escapeCharLiteral.getString(0).charAt(0) : DEFAULT_ESCAPE_CHAR;
```

Comment:
This will be a problem, because the size of the `ColumnVector` depends on how many rows the input has. For example if an empty batch is given, then this will fail with index out of bound exception as the `ColumnVector` will have zero rows.

One suggestion is to fetch it when it is needed. i.e when `getBoolean` on this method is called, initialize the escape character. At that time we know there the `ColumnVector` has at least one row.


```
public initEscapeCharIfRequired() {
   if (escapeChar == null) {
       escapeChar = escapeCharVector.getString(0).charAt(0) : DEFAULT_ESCAPE_CHAR;
   }
}
```

Then in the `boolen getBoolean(int rowId)`
```
   initEscapeCharIfRequired();
   isLike(left.getString(rowId), right.getString(rowId), escapeChar);
```

Reply:
There are otherways to handle this. When we resolve or validate the expression, we could create a custom Predicate with escape char as one of the class members.

```
public class ResolvedLike extends Predicate
{
   Expression rightSide;
   Expression leftSide,
   char escape;
}
```

We don't this now. I think the first solution has minimal code.

Reply:
you are spot on. I confirmed this with a test with empty batch. in fact the error from ColumnVector read max row index is -1. :)

Reply:
fixed.